xarray/tests/test_computation.py

```diff
@@ -789,6 +789,70 @@ def func(x):
     assert_identical(expected, actual)
 
 
+da = xr.DataArray(
+    np.random.random((3, 21, 4)),
+    coords={"time": pd.date_range("2000-01-01", freq="1D", periods=21)},
+    dims=("a", "time", "x"),
+)
+
+arrays = [
+    da.isel(time=range(0, 18)),
+    da.isel(time=range(2, 20)).rolling(time=3, center=True).mean(dim="time"),
+    xr.DataArray([0, np.nan, 1, 2, np.nan, 3, 4, 5, np.nan, 6, 7], dims="time"),
+    xr.DataArray([1, 1, np.nan, 2, np.nan, 3, 5, 4, 6, np.nan, 7], dims="time"),
+    xr.DataArray([[1, 2], [1, np.nan]], dims=["x", "time"]),
+    xr.DataArray([[1, 2], [np.nan, np.nan]], dims=["x", "time"]),
+]
+
+array_tuples = [
+    (arrays[0], arrays[0]),
+    (arrays[0], arrays[1]),
+    (arrays[1], arrays[1]),
+    (arrays[2], arrays[2]),
+    (arrays[2], arrays[3]),
+    (arrays[3], arrays[3]),
+    (arrays[4], arrays[4]),
+    (arrays[4], arrays[5]),
+    (arrays[5], arrays[5]),
+]
+
+
+@pytest.mark.parametrize("da_a, da_b", array_tuples)
+@pytest.mark.parametrize("dim", [None, "time"])
+def test_cov(da_a, da_b, dim):
+    def pandas_cov(ts1, ts2):
+        """Ensure the ts are aligned and missing values ignored"""
+        ts1, ts2 = xr.align(ts1, ts2)
+        valid_values = ts1.notnull() & ts2.notnull()
+
+        ts1 = ts1.where(valid_values, drop=True)
+        ts2 = ts2.where(valid_values, drop=True)
+
+        return ts1.to_series().cov(ts2.to_series())
```

Comment:
The assert works as it is right now, but let's keep in mind that `actual` is a `DataArray` and `expected` is a float (`np.float64`, I think).

I wrote a lengthy explanation on this in https://github.com/pydata/xarray/pull/3550#discussion_r348652667, but the short version is that I think it might make sense to construct a `DataArray` using the result from `pandas_cov` and to then use `assert_allclose` to compare.

Something like this, maybe?
```suggestion
        cov = ts1.to_series().cov(ts2.to_series())
        return xr.DataArray(cov)
```
though that won't solve the issues with the covariance along a dimension (which pandas does not support, I think), so I suspect you will have to rewrite `pandas_cov` and `pandas_corr` using numpy.

Reply:
You can also construct the result yourself—doing the 'compare with a known implementation' often saves time, but constructing the result is also fine!

Reply:
I changed to `xr.DataArray` and `assert_allclose`, but I will first stick with comparing to `pandas.Series` results.

I will rewrite `pandas_cov` and `pandas_corr`, do you have suggestions how to implement the `dim` parameter?

Reply:
:+1: for the `assert`, but let's take a look at the changes regarding `xr.DataArray`: I think you applied it in the wrong place. Since `cov` is a dataarray anyways, the result of `xr.cov` does not change (because `xr.DataArray(da).identical(da) == True` for any dataarray). The result of `pandas_cov` should be wrapped instead.

Reply:
For the result you could use `np.cov` and `np.corr`, but you would probably have to first transpose the second object so `dim` is in the same axis for both `DataArray` objects. Since `numpy` uses `axis` instead of `dim` you'd then determine the axis for `dim` (`DataArray.get_axis_num`, I think).

Reply:
However, that might be too much effort for a test and maybe not a good idea since you would end up reimplementing the function you want to test. So I think it probably would be better to follow @max-sixty's suggestion to construct the result yourself instead?

Reply:
Thanks for the feedback! 

Unfortunately, I am very busy these days, so I just added the xr.Dataarray() in a haste, hence the wrong placement...

I hope I can take some time the weekend on 7./8.12. Cheers!